tests/bson/bson-packedarray-fromJSON-003.phpt

```diff
@@ -0,0 +1,31 @@
+--TEST--
+MongoDB\BSON\PackedArray::fromJSON(): Decoding JSON object
+--FILE--
+<?php
+
+require_once __DIR__ . '/../utils/basic.inc';
+
+$bson = MongoDB\BSON\PackedArray::fromJSON('{ "foo": "bar" }');
+var_dump($bson);
+hex_dump((string) $bson);
+echo $bson->toCanonicalExtendedJSON(), PHP_EOL;
+echo MongoDB\BSON\Document::fromPHP(['array' => $bson])->toCanonicalExtendedJSON(), PHP_EOL;
```

Comment:
> Note that since `bson_new_from_json` was intended to read documents, the resulting PackedArray instance will retain the original keys in the object, but iterating it or converting it to JSON will reindex the array. However, when using the PackedArray instance in certain ways will cause errors.

I'm having trouble understanding why _this_ fails but the preceding `$bson->toCanonicalExtendedJSON()` does not.

Does `bson_json_opts_set_outermost_array()` apply some logic that disregards invalid keys, while the code path for converting a _nested_ array to extended JSON does not?

> The only option I see is to iterate the resulting bson_t and either re-index it (essentially re-writing a bson_t array) or to throw if one of the keys isn't what we expect. Since we have the mantra of "be lenient in what you accept, but strict in what you produce" I'm inclined to go with the first option, despite this potentially causing some overhead.

If [PackedArray::fromPHP()](https://www.php.net/manual/en/mongodb-bson-packedarray.fromphp.php) requires a well-formed list (see: [bson-packedarray-fromPHP_error-001.phpt](https://github.com/mongodb/mongo-php-driver/blob/v1.x/tests/bson/bson-packedarray-fromPHP_error-001.phpt)), I don't see why we can't require that in `fromJSON()` as well.

Perhaps we can't prevent libbson from accepting a JSON object with sequential numeric string keys starting at zero, but if it's easy to detect something like `{ "foo": "bar" }` I'm all for throwing an error there.

Reply:
I added logic to iterate the array and compare the keys. I've used `strtoll` to parse the string, as I didn't want to allocate memory to convert the expected key to a string. I'm not sure if that has any performance implications, but I don't expect this method to be somewhere in the hot path given its limited usefulness.

Reply:
Recapping our Slack discussion, we discussed a caveat with `strtoll()` accepting non-canonical numeric strings. I'll review the second approach.